File: CSharp/ExtensibleCommands/ExtensibleCommands/SimpleCommandIO.cs
using System;

namespace ExtensibleCommands
{
    /// <summary>
    /// Base class for atomic operation with input and output parameters
    /// </summary>
    /// <typeparam name="TInput">Type of input parameter</typeparam>
    /// <typeparam name="TOutput">Type of output parameter</typeparam>
    public class SimpleCommandIO<TInput, TOutput> : SimpleCommandI<TInput>
    {
        /// <summary> Type of delegate to execute a command with an input and an output parameters </summary>
        public delegate TOutput ExecutionDelegateIO(TInput input);

        /// <summary> Command output </summary>
        public TOutput Output { get; private set; }

        /// <summary> Delegate to execute </summary>
        protected new ExecutionDelegateIO _executionMethod;

        /// <summary> Constructor </summary>
        protected SimpleCommandIO()
        {
            Name = "Simple(Input, Output)";
        }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="executionDelegate">Delegate to execute</param>
        public SimpleCommandIO(ExecutionDelegateIO executionDelegate)
            : this(executionDelegate, "Simple(Input, Output)") { }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="executionDelegate">Delegate to execute</param>
        /// <param name="name"></param>
        public SimpleCommandIO(ExecutionDelegateIO executionDelegate, string name)
        {
            _executionMethod = executionDelegate;
            Name = name;
        }

        protected override void Execute()
        {
            if (_executionMethod != null)
                Output = _executionMethod.Invoke(Input);
        }
    }
}

File: Java/ExtensibleCommands/src/main/java/org/extensiblecommands/SimpleCommandIO.java
package org.extensiblecommands;

/**
 * Base class for atomic operation with input and output parameters
 * @param <TInput>      Type of input parameter
 * @param <TOutput>     Type of output parameter
 */
public class SimpleCommandIO<TInput, TOutput> extends SimpleCommandI<TInput> {
    /**
     * Command output
     */
    private TOutput output;

    /**
     * Delegate to execute
     */
    protected ExecutionDelegateIO<TInput, TOutput> executionMethod;

    /**
     * Constructor
     * @param executionDelegate     Delegate to execute
     */
    public SimpleCommandIO(ExecutionDelegateIO<TInput, TOutput> executionDelegate) {
        this(executionDelegate, "Simple(Input, Output)");
    }

    /**
     * Constructor
     * @param executionDelegate     Delegate to execute
     * @param name                  Command name
     */
    public SimpleCommandIO(ExecutionDelegateIO<TInput, TOutput> executionDelegate, String name) {
        executionMethod = executionDelegate;
        this.name = name;
    }

    /**
     * @return          Command output
     */
    public TOutput getOutput() {
        return output;
    }

    protected SimpleCommandIO() {
        this.name = "Simple(Input, Output)";
    }

    @Override
    protected void execute() throws Exception {
        if (executionMethod != null)
            this.output = executionMethod.execute(getInput());
    }
}
